config.go

```diff
@@ -0,0 +1,121 @@
+For example, to allocate and mount an audio capture device: {"name": "capture", "groups": [{"paths": [{"path": "/dev/snd/pcmC0D0c"}, {"path": "/dev/snd/controlC0"}]}]}
+For example, to expose a CH340 serial converter: {"name": "ch340", "groups": [{"usb": [{"vendor": "1a86", "product": "7523"}]}]}
+A "count" can be specified to allow a discovered device group to be scheduled multiple times.
+For example, to permit allocation of the FUSE device 10 times: {"name": "fuse", "groups": [{"count": 10, "paths": [{"path": "/dev/fuse"}]}]}
+Note: if omitted, "count" is assumed to be 1`)
+	flag.String("plugin-directory", v1beta1.DevicePluginPath, "The directory in which to create plugin sockets.")
+	flag.String("log-level", logLevelInfo, fmt.Sprintf("Log level to use. Possible values: %s", availableLogLevels))
+	flag.String("listen", ":8080", "The address at which to listen for health and metrics.")
+	flag.Bool("version", false, "Print version and exit")
+
+	flag.Parse()
+	if err := viper.BindPFlags(flag.CommandLine); err != nil {
+		return fmt.Errorf("failed to bind setup config: %w", err)
+	}
+
+	if *cfgFile != "" {
+		viper.SetConfigFile(*cfgFile)
+	} else {
+		viper.SetConfigName("config")
+		viper.SetConfigType("yaml")
+		viper.AddConfigPath("/config/")
+		viper.AddConfigPath("/etc/generic-device-plugin/")
+		viper.AddConfigPath(".")
+	}
+
+	viper.AutomaticEnv()
+	viper.SetEnvKeyReplacer(strings.NewReplacer(".", "_", "-", "_"))
+
+	if err := viper.ReadInConfig(); err != nil {
+		if _, ok := err.(viper.ConfigFileNotFoundError); ok {
+			// Config file not found; ignore error
+		} else {
+			// Config file was found but another error was produced
+			return fmt.Errorf("failed to read config file: %w", err)
+		}
+	}
+
+	viper.RegisterAlias("devices", "device")
+
+	return nil
+}
+
+// GetDevices returns a list of configured devices
+func GetDevices() ([]*deviceplugin.DeviceSpec, error) {
+	switch raw := viper.Get("device").(type) {
+	case []string:
+		// Assign deviceSpecs from flag
+		deviceSpecs := make([]*deviceplugin.DeviceSpec, len(raw))
+		for i, data := range raw {
+			if err := yaml.Unmarshal([]byte(data), &deviceSpecs[i]); err != nil {
+				return nil, fmt.Errorf("failed to parse device %q: %w", data, err)
+			}
+		}
+		return deviceSpecs, nil
+	case []interface{}:
+		// Assign deviceSpecs from config
+		deviceSpecs := make([]*deviceplugin.DeviceSpec, len(raw))
+		for i, data := range raw {
+			decoder, err := mapstructure.NewDecoder(&mapstructure.DecoderConfig{
+				Result:  &deviceSpecs[i],
+				TagName: "yaml",
+				DecodeHook: mapstructure.ComposeDecodeHookFunc(
+					deviceplugin.ToUSBIDHookFunc,
+				),
+			})
+			if err != nil {
+				return nil, err
+			}
+
+			if err := decoder.Decode(data); err != nil {
+				return nil, fmt.Errorf("failed to assign device %q: %w", data, err)
+			}
+		}
+		return deviceSpecs, nil
+	default:
+		return nil, fmt.Errorf("failed to fetch devices: unexpected type: %T", raw)
```

Comment:
Similarly, i think it would be clearer to say that we failed to decode the device.